Added support for Snowplow Analytics in AMP analytics suite (closes #1354)

alexanderdean opened the pull request:
Implements a minimal subset of the Snowplow Tracker Protocol (https://github.com/snowplow/snowplow/wiki/snowplow-tracker-protocol).

The user will need to set the following variables:

- 'host' to their collector endpoint, starting with http(s):// and *not* including a trailing slash
- 'appId' to identify their application ID

There is scope for adding more functionality in the future but this should get us started.

/cc @httpheaders, @luishgoncalves

cramforce commented:
Fixes #1354

This looks great!

@btownsend @avimehta Does this look good to you?

alexanderdean commented:
I will review & sign the CLA later tonight...

alexanderdean commented:
@fblundun can you sense check the PR on Monday?

As a note - I went with `'host'` because I saw that in the GA implementation; `'collectorUrl'` would be more idiomatic for existing Snowplow users...

cramforce commented:
Also CC @rudygalfi

fblundun commented:
@alexanderdean it looks good to me.

One question: are variables like `$screenWidth` guaranteed to be set? If not, what happens if they aren't? `res=${screenWidth}x${screenHeight}` becoming  `&res=undefinedxundefined` in the querystring will cause problems.

alexanderdean commented:
Good point @fblundun - question back to the AMP guys?

btownsend commented:
Re `host` vs `collectorUrl`, I'm not sure there's much value to copying `googleanalytics` attributes for the sake of consistency - the overall configuration will end up being very specific to each vendor. Choosing names that are familiar to your users is probably a better choice.

rudygalfi commented:
@fblundun I think when the variables aren't set you get an empty string, so `&res=x` in your example.

cc @btownsend

fblundun commented:
Thanks for that information @rudygalfi .

If screenWidth and screenHeight can be empty then we can change

```js
'res=${screenWidth}x${screenHeight}&stm=${timestamp}&'
```

to

```js
(screenWidth && screenHeight ? 'res=${screenWidth}x${screenHeight}&' : '') + 'stm=${timestamp}&'
```

cramforce commented:
We'd like the vendor configs to be declarative, so doing that string concat Kung-Fu would not be great.

I suppose we could add support for optional name value pairs, but I wonder whether that:

- is actually needed. Most servers should be fine with empty values
- in the example above: We should definitely always have screen dimensions.

alexanderdean commented:
Thanks for the feedback all. I have:

* Added `'transport': {'beacon': false, 'xhrpost': false, 'image': true},` as our `POST` behavior involves a different payload to our `GET`, so easiest just to stick to `GET`
* Replaced `${host}` with `${collectorUrl}` which is more idiomatic to Snowplow
* Left `'res=${screenWidth}x${screenHeight}&stm=${timestamp}&'` as they are

I will update this ticket once I've signed the CLA...

avimehta commented:
@alexanderdean When you are done, can you also update extensions/amp-analytics/amp-analytics.md?

rudygalfi commented:
I suggest making the update to extensions/amp-analytics/amp-analytics.md part of this pull request.

alexanderdean commented:
Done! Is there a plan for a more comprehensive manual which lists what vars you need to set for each event type? If you're planning to do that for GA, happy to copy/adapt it for Snowplow...

avimehta commented:
For now please add them to the same doc. If per-vendor stuff is moved to a
separate doc, we can do that for snowplow as well. (GA is just lagging
behind on documentation :) )

On Wed, Jan 13, 2016 at 2:23 PM, Alexander Dean <notifications@github.com>
wrote:

> Done! Is there a plan for a more comprehensive manual which lists what
> vars you need to set for each event type? If you're planning to do that for
> GA, happy to copy/adapt it for Snowplow...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amphtml/pull/1358#issuecomment-171455998>.
>

alexanderdean commented:
I signed the CLA! I believe everything is a wrap from our end - let me know if you need anything else...

rudygalfi commented:
cc @Meggin for docs review

Meggin commented:
lgtm... docs.

rudygalfi commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -63,6 +63,21 @@ export const ANALYTICS_CONFIG = {
+  'snowplow': {
+    'transport': {'beacon': false, 'xhrpost': false, 'image': true},
+    'requests': {
+      'aaVersion': 'amp-0.1',
```
Could `aaVersion` go into vars? Should it?

cc @avimehta

alexanderdean replied:
Your call but would make sense to me to have the tracker name/version available to all vendors...

avimehta replied:
I think it can. But I am not sure if others are going to use the same string/format. How about keeping it here for now and moving it to a shared string if others are going to need it?

alexanderdean replied:
Sounds good to me!

jmadler commented:
Is this an LGTM all-around, then? @avimehta, @cramforce

avimehta commented:
The commit needs an example in examples/analytics.amp.html and a merge from head to remove conflicts. Rest lgtm

avimehta commented:
Also, if possible, can you add the required vars to the doc? That way, people don't mistag the pages?

avimehta commented:
Any updates on this?

alexanderdean commented:
Sorry for delay @avimehta - that's been updated now!

rudygalfi commented:
Please rebase.

Also, can you add updates in these files as well?
* examples/analytics.amp.html
* extensions/amp-analytics/0.1/test/vendor-requests.json

See https://github.com/ampproject/amphtml/pull/2016/files for a sample.

alexanderdean commented:
Done!

rudygalfi commented:
@alexanderdean Thanks! Can you make changes to get Travis green? It looks like there's a comma-dangle error (might be others).

alexanderdean commented:
I am on it...

alexanderdean commented:
Fixed

rudygalfi commented:
@avimehta Can you give this one more pass since it's been awhile since the last reviews?

alexanderdean commented:
Any update on this one?

avimehta commented:
lgtm

avimehta commented:
@dvoytenko can you merge this one?
